tests/utils/state_locks_filter.py

```diff
@@ -0,0 +1,391 @@
+
+    def get_held_time(self) -> int:
+        return self.held_time
+
+    def get_wait_time(self) -> int:
+        return self.wait_time
+
+
+class LockTraceManager:
+    """
+    LockTraceManager: Handles filtering and managing the lock traces and associated group times.
+    It provides methods to filter the traces by time and to print the results in a sorted manner.
+    """
+
+    def __init__(self, traces: dict[LockOpTrace, list[GroupTimes]]):
+        self.traces = traces
+
+    # Filter the times for each trace
+    def filter(self, held_time: int, wait_time: int):
+        filtered_traces = dict[LockOpTrace, list[GroupTimes]]()
+        for trace, times in self.traces.items():
+            filtered_times = [
+                time
+                for time in self.traces.get(trace)
+                if time.get_held_time() >= held_time
+                and time.get_wait_time() >= wait_time
+            ]
+            if len(filtered_times) > 0:
+                filtered_traces[trace] = filtered_times
+
+        self.traces = filtered_traces
+
+    # Keep the traces that match with the params
+    def match(self, match_names: list[str]):
+        filtered_traces = dict[LockOpTrace, list[GroupTimes]]()
+        for trace, times in self.traces.items():
+            for match_name in match_names:
+                if trace.match(match_name):
+                    filtered_traces[trace] = times
+
+        self.traces = filtered_traces
+
+    # print the traces with their times for each test
+    def print(self, sort_held_time, sort_wait_time):
+        if sort_held_time:
+            for trace, times in self.traces.items():
+                self.traces[trace] = sorted(
+                    times, key=lambda x: x.get_held_time(), reverse=True
+                )
+        if sort_wait_time:
+            for trace, times in self.traces.items():
+                self.traces[trace] = sorted(
+                    times, key=lambda x: x.get_held_time(), reverse=True
+                )
+
+        for trace, times in self.traces.items():
+            print("-" * 20 + "TRACE" + "-" * 20)
+            print("")
+            print(trace)
+            print("")
+            for time in times:
+                print(
+                    "{} held: {} ms, wait: {} ms".format(
+                        time.get_group_name(),
+                        time.get_held_time(),
+                        time.get_wait_time(),
+                    )
+                )
+            print("")
+
+
+class LocksFileReader:
+    """
+    LocksFileReader: Reads the lock file and parses the different test cases and groups.
+    It extracts the relevant trace data and stores it in LocksGroup instances.
+    It can also return a dictionary of traces and associated group times.
+    """
+
+    PROJECT_PREFIX = "###START: SNAPD PROJECT"
+    TEST_PREFIX = "###START:"
+
+    lines: list[str]
+    groups: list[LocksGroup]
+
+    def __init__(self, locks_file: argparse.FileType):
```

Comment:
The type annotation is wrong, `argparse.FileType` is a function, not a type. You need `typing.IO[str]` here.